sqlite3/lib/src/impl/statement.dart

```diff
@@ -138,46 +168,75 @@ class PreparedStatementImpl implements PreparedStatement {
     for (var i = 1; i <= params.length; i++) {
       final Object? param = params[i - 1];
 
-      if (param == null) {
-        _bindings.sqlite3_bind_null(_stmt, i);
-      } else if (param is int) {
-        _bindings.sqlite3_bind_int64(_stmt, i, param);
-      } else if (param is double) {
-        _bindings.sqlite3_bind_double(_stmt, i, param.toDouble());
-      } else if (param is String) {
-        final bytes = utf8.encode(param);
-        final ptr = allocateBytes(bytes);
-        _allocatedWhileBinding.add(ptr);
+      _bindParam(param, i);
+    }
 
-        _bindings.sqlite3_bind_text(
-            _stmt, i, ptr.cast(), bytes.length, nullPtr());
-      } else if (param is List<int>) {
-        if (param.isEmpty) {
-          // malloc(0) is implementation-defined and might return a null
-          // pointer, which is not what we want: Passing a null-pointer to
-          // sqlite3_bind_blob will always bind NULL. So, we just pass 0x1 and
-          // set a length of 0
-          _bindings.sqlite3_bind_blob64(
-              _stmt, i, Pointer.fromAddress(1), param.length, nullPtr());
-        } else {
-          final ptr = allocateBytes(param).cast<Void>();
-
-          _bindings.sqlite3_bind_blob64(_stmt, i, ptr, param.length, nullPtr());
-          _allocatedWhileBinding.add(ptr);
-        }
-      } else {
+    _variablesBound = true;
+  }
+
+  void _bindMapParams(Map<String, Object?>? params) {
+    if (params == null || params.isEmpty) return;
+
+    // _bindings.sqlite3_clear_bindings(_stmt);
+
+    // variables in sqlite are 1-indexed
+    for (final key in params.keys) {
+      final Object? param = params[key];
+
+      final keyBytes = utf8.encode(key);
+      final keyPtr = allocateBytes(keyBytes, additionalLength: 1);
+      _allocatedWhileBinding.add(keyPtr);
+      final i = _bindings.sqlite3_bind_parameter_index(_stmt, keyPtr.cast());
+      if (i == 0) {
         throw ArgumentError.value(
-          param,
-          'params[$i]',
-          'Allowed parameters must either be null or an int, num, String or '
-              'List<int>.',
-        );
+            params, 'params', 'Key $key was not found in statement.');
       }
+      _bindParam(param, i);
     }
 
     _variablesBound = true;
```

Comment:
I started the mistake with `_ensureMatchingParameters`, but I think it would be easier if this method just checks the expected length. Then we don't have to call `_ensureMatchingMapParameters` at all use-sites and just have this check when binding parameters.

We should do the same thing with `_ensureMatchingParameters`, but I can take care of that too.